README
The APPLgrid project
====================

Introduction
------------

This contains the complete development installation of the 
APPLgrid project allowing the user to also generation grids
using custom versions of mcfm and nlojet++ which are included.


If simply performing the fast convolution is required, only 
the appl_grid directory is required, the hoppet installation 
if arbitrary scale changes during the fast convolution is 
required.

Installation prerequisites
--------------------------

Because the whole installation contains several distinct 
packages these is the 

 install.sh 

script, which can guide the installation, although the packages
can also be installed independently, although *please* ensure 
that your path contains instances of the 

 root-config
 lhapdf-config

package config utilities. In addition, if you have hoppet already 
installed, please ensure that 

 hoppet-config

is in your path.

In addition, the APPLgrid convolution code generates a config 
utility of it's own..

 applgrid-config

Please ensure that this is placed in your path also. 

All this can be ensured by specifying the usual 

  --prefix=<installation_base>

either to install.sh, or directly to the 

 ./configure  
 
script calls if building individually. 

In addition, please ensure that your LD_LIBRARY_PATH is set to include
the installation directory for all the dynamic libraries, and that 
if hoppet is installed independently it is installed with compile
option -fPIC for position independent code so it can be linked against 
dynamic library code. 

Installation
------------

By default, include files, libraries and the config utilities mentioned
above are installed in the include, lib and bin directories from the 
installation directory.

The nlojet executable is installed in the bin directory below the 
installation directory. 

A: Install and run the code

1) run the install script 

    install.sh

   the install script can be given arguments

   usage: install.sh [OPTIONS]

     --help| -h   this help
     --all|-a     do everything
     --none|-n    don't do anything
     --appl       install appl_grid
     --pdf        install hoppet
     --mcfm       install mcfm
     --nlo        install nlojet
     --mod        install nlojet module
     --runmcfm    run mcfm
     --runmod     run nlojet module

   The default is to install and run everything (except mcfm) so to 
   eg compile the appl_grid code only you should use

    install.sh --none --appl

   the argument --none being used to switch off everything before 
   enabling the appl_grid compilation.


Additional details, not checked, and which may be somewhat out of date,
but which may in any case, still be of help... 

1) run the setup.sh script (Only if you *really* need to, shouldn't any longer)
2) run the install.sh script. 
3) compile the fillgrid.cxx

   cd $BASEDIR/jetmod

   make fillgrid

4) Make the weighting tables with nlojet
   
   nlojet++ --calculate -c full -u fillgrid.la --max-event 1000010 
 
  (This command explores first the phase space to optimise the
   grid boundaries. In a next step the weighting tables are filled
   There is a problem, if you run to few events, then the program 
   never stops, since no weight file is produce. This we have to change...)

   The weighting tables are in the output directory 
   weight_c.root (central jets)
   weight_f.root (forward jets)

5) Example to convolute the weighting tables with the PDFs and alphas

   make stand

  ./stand output/weight_c.root

6) Check the result
   open root
   .x check.C
   to look at the histogram compring the results of the convolution
   of the grid with the reference histograms

B) Software organisation:

   files and directories in BASEDIRECTORY:
   README this files giving some basic explanations
   setup.sh: setup.sh file to set variables need to run nlojet and standalone
   install.sh: script to install all libraries and run some examples   


   appl_grid: The source code of the applgrid    

   jetmod: nlojet examples to make weight tables and validating them  

   lhpdf-1.0.0: The LHAPDF library interface for nlojet 

   mcfm:  The source library of the mcfm program                  

   nlojet++-4.0.1   The source library of the NLOJET progrom 
   bld-nlo: build directory of the nlojet program  
   bld-pdf: build directory of the nlojet pdf-code implementation  

C) appl_grid software:
   
   a) appl_grid.cxx  and appl_grid.h  (explained below)
   b) appl_igrid.cxx and appl_igrid.h (explained below)
   c) fappl_grid.cxx fortran callable wrapper 

   d) implementation of sparse matrix to store the weighting grid
   
      tsparse1d.h tsparse2d.h
      Sparse3d.h          tsparse3d.h
      sparse.h            tsparse_base.h
      SparseMatrix3d.cxx SparseMatrix3d.h

   e) pdf class, stores the combinations of the variuos pdf 
      elements to the more generalised subprocesses, ie 7 
      for the pp combinations, rather than the 13x13 

       base class: 
       appl_pdf.cxx and appl_pdf.h

       generalised pdf combinations for Z production: 
       mcfmz_pdf.h         
       mcfmz_pdf.cxx       

       for W production:
       mcfmw_pdf.h         
       mcfmw_pdf.cxx         

       and for jet production using either nlojet or jetrad:
       nlojet_pdf.cxx nlojet_pdf.h             
       jetrad_pdf.cxx  jetrad_pdf.h    

   g) helper functions to test CPU performance
       appl_timer.cxx    
       appl_timer.h    

   j) helper function: TObject string vector class for writing string vectors
       TFileString.cxx
       TFileString.h

      The following dictionary files are generated automatically
       TFileStringDict.h
       TFileStringDict.cxx 

Additional documentation (not checked)

        1) appl_grid.cxx  and appl_grid.h  (explained below)
           user interface for the grid software

        2) appl_igrid.cxx  and appl_igrid.h  (explained below)
           internal grid function where the action is

        3) constructors/distructors
        4) operators to add/copy grids

        5) void grid::fill_phasespace(const double x1, const double x2, const double Q2, 
                           const double obs, 
                           const double* weight, const int iorder)

        fill a default grid to explore the phase space in order to
        make a optimised grid in the next step

        6) void grid::trim() untrim() ??
   
        7) void grid::print() ??
           void grid::Write(const string& filename, const string& dirname)
           write the grid to a root file  

        8) void grid::pdfinterp(double x, double Q2, double* f) interpolate the pdfs
        9) convolute the grid with the pdf

        TH1D* grid::convolute(void (*pdf)(const double& , const double&, double* ), 
                      double (*alphas)(const double& ), 
                      int     nloops, 
                      double  rscale_factor,
                      double  fscale_factor,
                      void (*splitting)(const double& , const double&, double* ) )

        The use has to provide a pointer to the follwing functions:
        pdf       the interface to the pdf function
        alphas    the interface to the alphas function
        splitting the interface to the splitting function needed for the
                  factorisation scale dependence
        nloops =0 leading order  =1: next to leading order =-1: only NLO

        rscale_factor  renormalisation scale factor
        fscale_factor  factorsation    scale factor

        make a new operator in this function should be avoided

        10) TH1D* grid::convolute_subproc(int subproc,
                              void (*pdf)(const double& , const double&, double* ), 
                              double (*alphas)(const double& ), 
                              int     nloops, 
                              double  rscale_factor,
                              double  fscale_factor,
                              void (*splitting)(const double& , const double&, double* ) )
       as 9, but for each sub-process

       11) void grid::optimise() 
           void grid::optimise(int NQ2, int Nx)
           
           optimise the grid boundaries

       12) redefine grid boundaries by hand
           void grid::redefine(int iobs, int iorder,
                    int NQ2, double Q2min, double Q2max, 
                    int Nx,  double  xmin, double  xmax )
       

C) Some problems:

   1) on my ubunutu I had problems to run libtool in appl_grid
  
      I got
      libtool  X-tag=CXX command not found
 
       etc

       running autoreconf -i
       in ~/applgrid/appl_grid
       solved the problem


ChangeLog
2012-06-14  Gavin Salam  <salam@lpthe.jussieu.fr>

	* changed svn:external for hoppet to point to http://hoppet.hepforge.org/svn/trunk

2011-12-15 sutt -at- cern.ch
	* fix multiple scaleing of cross section with Escale for
	  subprocess contributions

2011-08-26 sutt -at- cern.ch
	* move mcfm customaisations from grid

2011-08-06 sutt -at- cern.ch
	* include mcfm 6.0 etc

2010-11-07  Gavin Salam  <salam@lpthe.jussieu.fr>

	* appl_grid/src/hoppet_init.cxx: 
	  added include <cstdlib>, for exit to be available on g++ 4.4

2009-08-16 sutt -at- cern.ch
	* first attempt at ChangeLog file
	
	


install.sh
#!/bin/sh 

######################################################
#  installation script - now greatly improved 
#  by default installs and runs everything 
#  except mcfm
#  
#  has control flags to allow any combination 
#  of jobs to be done. 
# 
#    APPLGRID - install appl_grid
#    PDF      - install hoppet
#    MCFM     - install mcfm
#    MCFM60   - install mcfm 6.0
#    RMCFM    - run mcfm
#    RMCFM60  - run mcfm 6.0
#    NLO      - install nlojet
#    NLOMOD   - instal nlojet module
#    RNLOMOD  - run nlojet module
#  
#  To do a single job, 
#  diasble everything with --none, ie 
#
#   install.sh --none --appl
#
######################################################


ARGS=$*


###############################################
#  setup everything
###############################################

if [ "$BASEDIR" = "" ]; then 
  export BASEDIR=$PWD
fi
export INSTALLBASE=$BASEDIR


# control flags

APPLGRID=1
PDF=1
MCFM=0
MCFM60=0
RMCFM=0
RMCFM60=0
NLO=1
NLOMOD=1
RNLOMOD=1
FASTJET=0
USERJ=1

#   echo "setting everything"
setall() { 
    APPLGRID=1
    PDF=1
    MCFM=1
    MCFM60=1
    RMCFM=1
    RMCFM60=1
    NLO=1
    NLOMOD=1
    RNLOMOD=1
    FASTJET=1
    USERJ=1
}

#   echo "unsetting everything"
unsetall() { 
    APPLGRID=0
    PDF=0
    MCFM=0
    MCFM60=0
    RMCFM=0
    RMCFM60=0
    NLO=0
    NLOMOD=0
    RNLOMOD=0
    FASTJET=0
    USERJ=0
}

# usage message
usage() { 
    echo "usage: install.sh [OPTIONS]\n"
    echo "  --help| -h   this help"
    echo 
    echo "  --all|-a        do everything"
    echo "  --none|-n       don't do anything"
    echo "  --appl          install appl_grid"
    echo "  --pdf|--hoppet  install hoppet"
    echo "  --mcfm          install mcfm"
    echo "  --mcfm60        install mcfm v6.0"
    echo "  --nlo           install nlojet"
    echo "  --mod           install nlojet module"
    echo
    echo "  --runmcfm       run mcfm"
    echo "  --runmcfm60     run mcfm v6.0"
    echo "  --runmod        run nlojet module"
    echo 
    echo "  --user          build and run simple user example"
    echo "  --fastjet       build fastjet"
    echo
    echo "  --prefix=<directory>  set the installation directory"
    echo 
    echo "\nReport bugs to <sutt@cern.ch>" 
#   exit
}


# set the installation diectory if required
prefix() {
    export INSTALLBASE=`echo $PREFIX | sed "s#--prefix=##"`
#   echo installing in $INSTALLBASE
} 



###############################################
# parse arguments and set control flags 
###############################################


for WORD in $ARGS ; do
   case $WORD in
       --all|-a)   setall;;  
       --none|-n)  unsetall;;
       --appl)     APPLGRID=1;;
       --pdf|--hoppet)      PDF=1;;
       --mcfm)     MCFM=1;;
       --mcfm60)   MCFM60=1;;
       --runmcfm)  RMCFM=1;;
       --runmcfm60) RMCFM60=1;;
       --nlo)      NLO=1;;
       --mod)      NLOMOD=1;;
       --runmod)   RNLOMOD=1;;
       --user)     USERJ=1;;
       --fastjet)  FASTJET=1;;
       --prefix=*) PREFIX=$WORD;prefix;;
       --help|-h)  usage;exit;;
       *)          usage;exit;;
   esac
done


echo "base directory " $BASEDIR 
if [ -e $INSTALLBASE ]; then 
   echo "installing in $INSTALLBASE"
else
   echo "$INSTALLBASE does not exist" 
   exit
fi


cd   $BASEDIR

export PATH=$INSTALLBASE/bin:$BASEDIR/bin:$PATH

ROOTFOUND=`which root-config`

if [ "$ROOTFOUND" = "" ]; then 
  echo "root not setup"
  exit
fi

LHAPDFFOUND=`which lhapdf-config`
 
if [ "$LHAPDFFOUND" = "" ]; then 
  echo "lhapdf not setup"
  exit
fi



#  use root-config to decide what architecture root libraries will be used
 
# export ARCH=`rootarch.sh`
export ARCH=`root-config --ldflags`
export SYSARCH=`setarch.sh`

export CXXFLAGS=$ARCH
export F90FLAGS=$ARCH
export F77FLAGS=$ARCH
export CFLAGS=$ARCH
export FC=gfortran
# export FFLAGS="$ARCH -O"
export FFLAGS=$ARCH
export LDFLAGS=$ARCH

export HOPPETLIBS=
export HOPPETINCS=
export HOPPETFLAG=  

export FASTJETLIBS=
export FASTJETINCS=
export FASTJETFLAG=  




export ROOTSYS=`root-config --prefix`
export LHAPDFLIB=`lhapdf-config --prefix`/lib


export PATH=$ROOTSYS/bin:$PATH
export LD_LIBRARY_PATH=$ROOTSYS/lib:$LD_LIBRARY_PATH
export LD_LIBRARY_PATH=$INSTALLBASE/lib:$INSTALLBASE/libexec:$LHAPDFLIB:$LD_LIBRARY_PATH
export DYLD_LIBRARY_PATH=$INSTALLBASE/lib:$INSTALLBASE/libexec:$LHAPDFLIB:$DYLD_LIBRARY_PATH




############################
# INSTALL THE APPL GRID CODE 
############################
install_appl_grid() { 

    cd $BASEDIR/appl_grid

    ./configure --prefix=$INSTALLBASE

    #    if [ "$1" = "clean" ]; then
    #	   make clean
    #    fi

    make 
    make install

    hash -r

}



############################
# INSTALL EVOLUTION CODE
############################
install_pdf() { 

    cd $BASEDIR/hoppet

    if [ -e $BASEDIR/hoppet/benchmarking/benchmark.f90 ]; then 
         rm $BASEDIR/hoppet/benchmarking/benchmark.f90 
    fi

    if [ "$1" = "clean" ]; then
	make clean
    fi
    ./configure --prefix=$INSTALLBASE FC=$FC FFLAGS=" -fPIC $FFLAGS" LDFLAGS="$LDFLAGS"
    make
#   make check
    make install

    hash -r

    # cd $BASEDIR/pdf-conv-1.0/src
    # 
    # if [ "$1" = "clean" ]; then
    # 	make clean
    # fi
    # 
    # make  install
    
}



######################
# INSTALL MCFM
######################
install_mcfm() { 
    cd $BASEDIR
    cd $BASEDIR/mcfm
    make install    
}

install_mcfm60() {
    cd $BASEDIR
    cd $BASEDIR/mcfm-6.0
    make install    
#   should no longer be needed by lhapdf. hmmmm
    cd Bin
    if [ -e PDFsets ]; then 
	ls -ld PDFsets
    else
	ln -s `lhapdf-config --pdfsets-path` .
	ls -ld PDFsets
    fi
    

}


############################################
# RUN THE MCFM TEST EXECUTABLE
###########################cd ..#################
run_mcfm() { 

  if [ -d $BASEDIR/mcfm/run ]; then 

    cd $BASEDIR/mcfm/run
    rm -f *.log

#   should no longer be needed by lhapdf. hmmmm
    if [ -e PDFsets ]; then 
	ls -ld PDFsets
    else
	ln -s `lhapdf-config --pdfsets-path` .
	ls -ld PDFsets
    fi
    
    echo "running mcfm - please wait ..."

    if [ -e ../exe/$SYSARCH/mcfm ]; then 

       rm -rf W_*	

       #### Wminus test executable ####
#      ../exe/$SYSARCH/mcfm WMinput.DAT >&  mcfm-wm0.log
#      ../exe/$SYSARCH/mcfm WMinput.DAT >&  mcfm-wm1.log
#      ../exe/$SYSARCH/stand grid-30-Wminus_eta4.root
#      mv fout.root fout-Wminus.root

      rm -rf W_*	

#     #### Wplus test executable ####
      ../exe/$SYSARCH/mcfm WPinput.DAT >&  mcfm-wp0.log
      ../exe/$SYSARCH/mcfm WPinput.DAT >&  mcfm-wp1.log
      ../exe/$SYSARCH/stand grid-30-Wplus_eta4.root
      mv fout.root fout-Wplus.root

      cd ..
    fi
  fi
}


run_mcfm60() { 

  if [ -d $BASEDIR/mcfm-6.0/Bin ]; then 

    cd $BASEDIR/mcfm-6.0/Bin
    rm -f *.log

#   should no longer be needed by lhapdf. hmmmm
    if [ -e PDFsets ]; then 
	ls -ld PDFsets
    else
	ln -s `lhapdf-config --pdfsets-path` .
	ls -ld PDFsets
    fi
    
    echo "running mcfm - please wait ..."

    if [ -e mcfm ]; then 

      rm -rf W_*	

#     #### Wplus test executable ####
      mcfm winput.DAT >&  mcfm.log
      mcfm winput.DAT >&  mcfm.log
      stand grid-30-Wplus_eta4.root
      mv fout.root fout-Wplus.root

      cd ..
    fi
  fi
}


install_nlojet() { 
#############################
# INSTALL NLOJET PDF LIBRARY 
#############################
    
    cd $BASEDIR
    
    echo  "bld-pdf" 
    
    if [ "$1" = "clean" ]; then
	rm -rf bld-pdf
    fi
    
    if [ "$1" = "nloclean" ]; then
	rm -rf bld-pdf
    fi
    
    
    if [ -e bld-pdf ]; then 
	echo "bld-pdf already exists" 
    else
	
	mkdir bld-pdf
	
	if [ "$1" = "clean" ]; then
	    make clean
	fi
	
	cd bld-pdf
	../lhpdf-1.0.0/configure --prefix=$INSTALLBASE
	
	make        
	make install 

    fi

    hash -r


######################
# INSTALL NLOJET
######################
    
    cd $BASEDIR
    
    
    echo "bld-nlo" 
    if [ -e bld-nlo ]; then 
	echo "bld-nlo already exists" 
    else
	mkdir bld-nlo
	
	cd   bld-nlo
	
	echo "nlojet"
	../nlojet++-4.0.1/configure --prefix=$INSTALLBASE 
	
	if [ "$1" = "clean" ]; then
	    make clean
	fi
	
	make          
	make install 
	
    fi
    
    hash -r

    echo $PATH
    echo "which create-nlojet-user"

#   which create-nlojet-user
    
}




######################################
# INSTALL NLOJET GRID FILLING MODULES 
######################################
install_nlojet_module() { 

    echo "jetmod"
    if [ -e $BASEDIR/jetmod  ]; then
	echo "cd jetmod"
	cd $BASEDIR/jetmod
    else 
	echo "$BASEDIR/jetmod doesn't exist"
	exit
    fi
    

    echo "building fillgrid..."
    
    if [ "$1" = "clean" ]; then
	make clean
    fi
    
    make install
    
    hash -r

    echo PATH is: $PATH
    echo LD_LIBRARY_PATH: $LD_LIBRARY_PATH
 
    # recent LHAPDF should no longer need this   
    if [ -e PDFsets ]; then 
	ls -ld PDFsets
    else
	ln -s `lhapdf-config --pdfsets-path` .
	ls -ld PDFsets
    fi
    
}


##################################
# RUN NLOJET GRID FILLING MODULES 
##################################
run_nlojet_module () { 

    cd $BASEDIR/jetmod

    if [ -e output/weight_c.root ]; then
	echo "nlojet already executed"
    else  
	which nlojet++
#       nlojet++ --calculate -u fillgrid.la --max-event 20000
#       generate the grid phase space
        nlojet++ --calculate -c full -u fillgrid.la --max-event 1000010
#       now optimise it and generate the weights properly
        nlojet++ --calculate -c full -u fillgrid.la --max-event 1000010
    fi

    if [ -e output/weight_c.root ]; then 
#	./standalone output/weight_c.root
	./stand      output/weight_c.root
    else
	echo "nlojet grid code did not complete correctly" 
    fi
    
}


run_user () { 
    cd $BASEDIR/user

    if [ "$1" = "clean" ]; then
	make clean
    fi

    make all

    echo "\nrunning executables..."

    if [ -e ../jetmod/output/weight_c.root ]; then 
      ./stand ../jetmod/output/weight_c.root
    else
       echo "weight file not found (../jetmod/output/weight_c.root)"
       echo "have you already run nlojet++ ?"
    fi
       
    if [ -e fnl0004.tab ]; then 
      ./fnlo  fnl0004.tab
    else 
      echo "fastnlo grid file (fnl0004.tab) missing" 	
    fi

}

install_fastjet () { 

    cd $BASEDIR

    if [ -e "$BASEDIR/fastjet-*" ]; then
       echo "fastjet already installed"
    else 
      FASTJETVER=`wget --quiet http://www.lpthe.jussieu.fr/~salam/fastjet/version.html -O - | grep '^[0-9]' -` 
      curl http://www.lpthe.jussieu.fr/~salam/fastjet/repo/fastjet-$FASTJETVER.tar.gz > /tmp/fastjet-$FASTJETVER.tar.gz
      tar -xzf /tmp/fastjet-$FASTJETVER.tar.gz 

      if [ -e "$BASEDIR/fastjet-$FASTJETVER" ]; then 
         cd $BASEDIR/fastjet-$FASTJETVER
         ./configure --prefix=$INSTALLBASE  FC=$FC CXXFLAGS="$CXXFLAGS" FFLAGS="$FFLAGS" CFLAGS="$CFLAGS" LDFLAGS="$LDFLAGS"
         if [ "$1" = "clean" ]; then
 	   make clean
         fi
         make install
      fi

    fi

}


###########################
# ACTUALLY RUN THE STAGES 
###########################

hash -r

if [ "$PDF" = 1 ];      then install_pdf        $ARGS; fi

if [ -e $INSTALLBASE/bin/hoppet-config ]; then 
    export HOPPETLIBS=` $INSTALLBASE/bin/hoppet-config --libs `
    export HOPPETINCS=` $INSTALLBASE/bin/hoppet-config --cxxflags ` 
    export HOPPETFLAG="  -DHOPPET " 
fi 

if [ "$APPLGRID" = 1 ]; then install_appl_grid  $ARGS; fi
if [ "$MCFM" = 1 ];     then install_mcfm       $ARGS; fi
if [ "$MCFM60" = 1 ];   then install_mcfm60     $ARGS; fi
if [ "$RMCFM" = 1 ];    then run_mcfm;                 fi        
if [ "$RMCFM60" = 1 ];  then run_mcfm60;               fi        
if [ "$NLO" = 1 ];      then install_nlojet     $ARGS; fi


FASTJETFOUND=`which fastjet-config`

if [ "$FASTJETFOUND" == "" ]; then  
  if [ "$FASTJET" = 1 ];  then    install_fastjet        $ARGS; fi
 
  if [ -e $INSTALLBASE/bin/fastjet-config ]; then 
     export FASTJETLIBS=` $INSTALLBASE/bin/fastjet-config --libs `
     export FASTJETINCS=` $INSTALLBASE/bin/fastjet-config --cxxflags ` 
     export FASTJETFLAG="  -DFASTJET " 
  fi
else
  if [ -e $INSTALLBASE/bin/fastjet-config ]; then 
     export FASTJETLIBS=` fastjet-config --libs `
     export FASTJETINCS=` fastjet-config --cxxflags ` 
     export FASTJETFLAG="  -DFASTJET " 
  fi
fi 

if [ "$NLOMOD" = 1 ];   then    install_nlojet_module  $ARGS; fi
if [ "$RNLOMOD" = 1 ];  then    run_nlojet_module;            fi 
if [ "$USERJ" = 1 ];    then    run_user           $ARGS;  fi 
